File: rust/tcl-bigip/src/flow/sessions.rs
//! Flow pairing: directional flows → Connections → client/server Sessions.

use std::collections::HashMap;

use indexmap::IndexMap;

use super::model::{Connection, Flow, FlowKey, Session};

/// Pair opposite-direction flows into bidirectional [`Connection`]s.
///
/// The SYN-bearer is preferred as the client side. Each flow appears in exactly
/// one connection; orphan flows (no reverse seen) get `server = None`.
#[must_use]
pub fn pair_connections(flows: &IndexMap<FlowKey, Flow>) -> Vec<Connection> {
    let mut remaining: HashMap<FlowKey, Flow> =
        flows.iter().map(|(k, v)| (k.clone(), v.clone())).collect();

    // Stable sort over the insertion-ordered keys by (syn-rank, -packets).
    let syn_rank = |f: &Flow| -> u8 {
        if f.tcp_syn {
            0
        } else if f.tcp_synack {
            1
        } else {
            2
        }
    };
    let mut keys_in_order: Vec<FlowKey> = flows.keys().cloned().collect();
    keys_in_order.sort_by(|a, b| {
        let fa = &flows[a];
        let fb = &flows[b];
        syn_rank(fa)
            .cmp(&syn_rank(fb))
            .then_with(|| fb.packets.cmp(&fa.packets))
    });

    let mut out: Vec<Connection> = Vec::new();
    for key in keys_in_order {
        let Some(mut flow) = remaining.remove(&key) else {
            continue;
        };
        let rev: FlowKey = (
            flow.dst_ip.clone(),
            flow.dst_port,
            flow.src_ip.clone(),
            flow.src_port,
            flow.proto,
        );
        let mut peer = remaining.remove(&rev);
        if let Some(p) = &peer
            && !flow.tcp_syn
            && p.tcp_syn
        {
            std::mem::swap(&mut flow, peer.as_mut().unwrap());
        }
        out.push(Connection {
            client: flow,
            server: peer,
        });
    }
    out
}

/// Pair flows into Connections, then pair Connections into Sessions.
///
/// On a `tcpdump -i <vlan>:np` capture every TMM-mediated packet appears twice,
/// each carrying the proxied peer-side 5-tuple in its F5 ethernet trailer. The
/// front-side client flow's `peer_*` fields point at the back-side connection.
#[must_use]
pub fn pair_sessions(flows: &IndexMap<FlowKey, Flow>) -> Vec<Session> {
    let conns = pair_connections(flows);

    // Index every connection by the 5-tuple of its client flow.
    let mut by_client_key: HashMap<FlowKey, usize> = HashMap::new();
    for (i, c) in conns.iter().enumerate() {
        by_client_key.insert(c.client.key(), i);
    }

    let mut used: Vec<bool> = vec![false; conns.len()];
    let mut sessions: Vec<Session> = Vec::new();
    for i in 0..conns.len() {
        if used[i] {
            continue;
        }
        let f = &conns[i].client;
        if !f.peer_remote_ip.is_empty()
            && f.peer_remote_port != 0
            && !f.peer_local_ip.is_empty()
            && f.peer_local_port != 0
        {
            let back_key: FlowKey = (
                f.peer_local_ip.clone(),
                f.peer_local_port,
                f.peer_remote_ip.clone(),
                f.peer_remote_port,
                f.proto,
            );
            if let Some(&j) = by_client_key.get(&back_key)
                && j != i
            {
                used[i] = true;
                used[j] = true;
                sessions.push(Session {
                    front: conns[i].clone(),
                    back: Some(conns[j].clone()),
                });
                continue;
            }
        }
        used[i] = true;
        sessions.push(Session {
            front: conns[i].clone(),
            back: None,
        });
    }
    sessions
}

#[cfg(test)]
mod tests {
    use indexmap::IndexMap;

    use super::super::model::{Flow, FlowKey};
    use super::{pair_connections, pair_sessions};

    fn flow(src: &str, sp: u16, dst: &str, dp: u16) -> Flow {
        Flow::new(src.into(), sp, dst.into(), dp, 6)
    }

    #[test]
    fn pair_connections_pairs_opposite_flows_with_syn_bearer_as_client() {
        let mut client = flow("10.0.0.1", 1000, "10.0.0.2", 80);
        client.tcp_syn = true;
        client.packets = 5;
        let mut server = flow("10.0.0.2", 80, "10.0.0.1", 1000);
        server.tcp_synack = true;
        server.packets = 4;

        let mut flows: IndexMap<FlowKey, Flow> = IndexMap::new();
        // Insert the server side first to show the SYN-bearer still wins.
        flows.insert(server.key(), server.clone());
        flows.insert(client.key(), client.clone());

        let conns = pair_connections(&flows);
        assert_eq!(conns.len(), 1);
        assert_eq!(
            conns[0].client.src_port, 1000,
            "the SYN-bearer is the client"
        );
        assert_eq!(
            conns[0].server.as_ref().map(|s| s.src_port),
            Some(80),
            "the reverse flow becomes the server"
        );
    }

    #[test]
    fn pair_connections_leaves_orphans_without_a_server() {
        let mut only = flow("10.0.0.1", 1000, "10.0.0.2", 80);
        only.tcp_syn = true;
        let mut flows: IndexMap<FlowKey, Flow> = IndexMap::new();
        flows.insert(only.key(), only.clone());
        let conns = pair_connections(&flows);
        assert_eq!(conns.len(), 1);
        assert!(
            conns[0].server.is_none(),
            "no reverse flow → server is None"
        );
    }

    #[test]
    fn pair_sessions_links_front_and_back_via_peer_trailer() {
        // A front-side client flow whose F5 trailer peer-tuple points at the
        // back-side connection's client 5-tuple.
        let mut front = flow("10.0.0.1", 1000, "10.0.0.2", 80);
        front.tcp_syn = true;
        front.peer_local_ip = "172.16.0.1".into();
        front.peer_local_port = 5000;
        front.peer_remote_ip = "192.168.1.1".into();
        front.peer_remote_port = 443;
        // The back connection's client flow, keyed exactly by that peer-tuple.
        let mut back = flow("172.16.0.1", 5000, "192.168.1.1", 443);
        back.tcp_syn = true;

        let mut flows: IndexMap<FlowKey, Flow> = IndexMap::new();
        flows.insert(front.key(), front.clone());
        flows.insert(back.key(), back.clone());

        let sessions = pair_sessions(&flows);
        assert_eq!(sessions.len(), 1, "front and back fold into one session");
        let s = &sessions[0];
        assert_eq!(s.front.client.dst_port, 80);
        let back_conn = s.back.as_ref().expect("back side is linked");
        assert_eq!(back_conn.client.src_ip, "172.16.0.1");
    }

    #[test]
    fn pair_sessions_without_peer_info_has_no_back() {
        let mut lone = flow("10.0.0.1", 1000, "10.0.0.2", 80);
        lone.tcp_syn = true;
        let mut flows: IndexMap<FlowKey, Flow> = IndexMap::new();
        flows.insert(lone.key(), lone.clone());
        let sessions = pair_sessions(&flows);
        assert_eq!(sessions.len(), 1);
        assert!(sessions[0].back.is_none());
    }
}

File: dialects/f5/bigip/flow/sessions.py
# tcl-lsp — a language server and toolchain for Tcl
# Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: AGPL-3.0-or-later

"""Flow pairing: directional flows -> Connections -> client/server Sessions."""

from __future__ import annotations

from ._model import Connection, Flow, Session


def pair_connections(flows: dict[tuple[str, int, str, int, int], Flow]) -> list[Connection]:
    """Pair opposite-direction flows into bidirectional :class:`Connection`s.

    The SYN-bearer is preferred as the client side.  Each flow appears
    in exactly one connection; orphan flows (no reverse seen) are
    emitted as a connection with ``server=None``.
    """
    remaining = dict(flows)
    out: list[Connection] = []
    keys_in_order = sorted(
        remaining.keys(),
        key=lambda k: (
            0 if remaining[k].tcp_syn else (1 if remaining[k].tcp_synack else 2),
            -remaining[k].packets,
        ),
    )
    for key in keys_in_order:
        if key not in remaining:
            continue
        flow = remaining.pop(key)
        rev = (flow.dst_ip, flow.dst_port, flow.src_ip, flow.src_port, flow.proto)
        peer = remaining.pop(rev, None)
        if peer is not None and not flow.tcp_syn and peer.tcp_syn:
            flow, peer = peer, flow
        out.append(Connection(client=flow, server=peer))
    return out


def pair_sessions(
    flows: dict[tuple[str, int, str, int, int], Flow],
) -> list[Session]:
    """Pair flows into Connections, then pair Connections into Sessions.

    On a `tcpdump -i <vlan>:np` capture every TMM-mediated packet
    appears twice — once on the front (client-facing) side and once on
    the back (pool-member-facing) side.  The F5 ethernet trailer's
    HIGH TLV carries the proxied peer-side 5-tuple, so we can match a
    front-side Connection ``(client_ip:cport <-> vip:vport)`` with the
    back-side Connection it generated ``(snat_ip:sport <-> member_ip:mport)``
    by looking at the front-side client flow's
    ``(peer_remote_ip, peer_remote_port, peer_local_ip, peer_local_port)``
    values.

    Sessions emit only-front when no peer info is present (single-side
    capture) and only-back if a back-side Connection exists with no
    matching front (rare; usually means the front got dropped before
    the trailer was emitted).
    """
    conns = pair_connections(flows)

    # Index every connection by the 5-tuple of its client flow so we can
    # look up the back-side via the peer info on the front-side client flow.
    by_client_key: dict[tuple[str, int, str, int, int], Connection] = {}
    for c in conns:
        by_client_key[c.client.key] = c

    used: set[int] = set()
    sessions: list[Session] = []
    for c in conns:
        if id(c) in used:
            continue
        # Try treating *c* as the front side; look up the back-side
        # connection whose client 5-tuple matches the front client's
        # peer tuple.
        f = c.client
        if f.peer_remote_ip and f.peer_remote_port and f.peer_local_ip and f.peer_local_port:
            # Front-side client → VIP, peer = local→remote on the back side
            # (TMM as the proxied client, member as the proxied server).
            back_key = (
                f.peer_local_ip,
                f.peer_local_port,
                f.peer_remote_ip,
                f.peer_remote_port,
                f.proto,
            )
            back = by_client_key.get(back_key)
            if back is not None and id(back) != id(c):
                used.add(id(c))
                used.add(id(back))
                sessions.append(Session(front=c, back=back))
                continue
        used.add(id(c))
        sessions.append(Session(front=c, back=None))
    return sessions


# ---------------------------------------------------------------------------
# Optional tshark enrichment.
# ---------------------------------------------------------------------------
